### PR TITLE
Add @RCGitBot please test <job-name> on-demand job trigger

### PR DESCRIPTION
**Pull request by ajpallares**
## Summary

Extend the dynamic config setup to support running specific CircleCI jobs on-demand from PR comments. Multiple jobs can be requested in a single command, separated by spaces.

## How it works

The GitHub Actions workflow `rcgitbot_please_test.yml` handles two distinct commands via separate GHA jobs:

- **`@RCGitBot please test`** (exact match, no job names) → triggers the `approve-full-tests` GHA job, which approves the CircleCI hold job to run the full test suite. No new pipeline is created.
- **`@RCGitBot please test <job-names>`** (with job names after the prefix) → triggers the `run-specific-jobs` GHA job, which calls the CircleCI API with a `requested_jobs` parameter to create a new pipeline. The setup config then runs `generate-requested-jobs-config.js` to produce a continuation config containing only the requested CircleCI jobs.

Examples:
- `@RCGitBot please test run-test-ios-26`
- `@RCGitBot please test backend-integration-tests-SK1 backend-integration-tests-SK2`
- `@RCGitBot please test lint api-tests pod-lib-lint`

The `generate-requested-jobs-config.js` script:
1. Validates each requested job against an allowlist (release/deployment jobs are excluded). Each allowlisted job is mapped to the CircleCI contexts it needs, matching what's declared in `default_config.yml`.
2. Copies all orbs, executors, commands, and job definitions from `default_config.yml`
3. Appends a single `on-demand-jobs` workflow containing only the requested CircleCI jobs, with their per-job contexts
4. Continues the pipeline with this generated config

## Changes

- **`.circleci/config.yml`**: Add `requested_jobs` string parameter. When set, the setup phase generates a continuation config with only the requested jobs. Normal pipelines are unaffected. Setup job now uses `cimg/node:current` since the script is Node-based.
- **`.circleci/default_config.yml`**: Add `requested_jobs` parameter declaration (required for implicit forwarding).
- **`.circleci/generate-requested-jobs-config.js`**: External Node script that validates jobs against an allowlist (with per-job contexts) and generates the continuation config.
- **`.github/workflows/rcgitbot_please_test.yml`** (renamed from `approve_full_tests.yml`): Add `run-specific-jobs` GHA job that handles `@RCGitBot please test <job-names>` comments — verifies org membership, parses job name(s), triggers a CircleCI pipeline via API, and posts a link.

## Dependencies

- Depends on #6605 (dynamic config migration)

## Security

- Only RevenueCat org members can trigger jobs (verified via GitHub API)
- Job names are validated against `[a-zA-Z0-9_-]` regex before being passed to CircleCI
- Jobs are validated against an allowlist — release/deployment CircleCI jobs cannot be triggered this way

## Test plan

- [x] Comment `@RCGitBot please test` on a PR → should still approve the full test suite (no regression)
- [x] Trigger multiple jobs via CircleCI UI with `requested_jobs` parameter
- [ ] Comment `@RCGitBot please test run-test-ios-26` on a PR → should trigger only that CircleCI job
- [ ] Comment `@RCGitBot please test nonexistent-job` → should fail with an error
- [ ] Comment from a non-org member → should be rejected

---

> [!NOTE]
> **Medium Risk**
> Touches CI orchestration and adds an API-triggered path that runs jobs with contexts/secrets; correctness of the allowlist, parsing, and continuation config generation is key to avoiding unintended job execution or secret exposure.
> 
> **Overview**
> Adds an on-demand CI path that lets RevenueCat org members comment `@RCGitBot please test <job...>` to trigger a new CircleCI pipeline that runs only the requested allowlisted job(s).
> 
> This introduces a new `requested_jobs` pipeline parameter and setup-phase branching in `.circleci/config.yml` to generate and continue with a minimal workflow config, backed by a new Node script (`.circleci/generate-requested-jobs-config.js`) that validates job names and applies required contexts. The GitHub Action workflow is extended with a `run-specific-jobs` job that verifies org membership, parses/validates job names, calls the CircleCI API with `requested_jobs`, and posts a pipeline link back to the PR comment.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5f92610d8283ce758ec681b4045971d3ffb6d6c8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by ajpallares** [comment]
@RCGitBot please test

**Comment by ajpallares** [comment]
@RCGitBot please test run-test-ios-26

File: .circleci/config.yml [hunk 1]
```diff
@@ -29,6 +29,11 @@ parameters:
   generate_swiftinterface:
     type: boolean
     default: false
+  # Set via CircleCI API by the GitHub Action when a PR comment like
+  # "@RCGitBot please test <job1> <job2>, <job3>" is posted.
```
Comment (tonidero) [inline .circleci/config.yml]: Hmm the commas are confusing... does it require them?

File: .circleci/generate-requested-jobs-config.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+head -n "$WF_LINE" "$CONFIG" > /tmp/requested-jobs-config.yml
+
+{
+  echo "  on-demand-jobs:"
+  echo "    jobs:"
+  for JOB in $JOBS; do
+    echo "      - ${JOB}:"
+    echo "          context:"
+    echo "            - e2e-tests"
+    echo "            - slack-secrets"
```
Comment (tonidero) [inline .circleci/generate-requested-jobs-config.sh]: Hmm I guess this could potentially be adding secrets unnecessarily to some jobs 🤔. I wonder if this should be some sort of map, to know which contexts are required per job? Also, this might be easier to type with typescript/javascript, even though it would require a node machine.

File: .circleci/config.yml [hunk 3]
```diff
@@ -50,8 +55,19 @@ jobs:
...
     steps:
       - checkout
-      - continuation/continue:
-          configuration_path: .circleci/default_config.yml
+      - when:
+          condition: << pipeline.parameters.requested_jobs >>
+          steps:
+            - run:
+                name: Generate requested jobs configuration
+                command: bash .circleci/generate-requested-jobs-config.sh "<< pipeline.parameters.requested_jobs >>"
```
Comment (tonidero) [inline .circleci/config.yml]: I wonder if this whole job should be part of the orb... But we can keep it here for now since I think for now we might only be adding this to iOS, which will allow faster iteration, and then we can think about moving to the orb in the future if needed 👍 .

File: .circleci/generate-requested-jobs-config.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  run-test-ios-26
+  run-test-tvos-and-macos
+  run-test-watchos
+  spm-receipt-parser
+  spm-revenuecat-ui-ios-15
+  spm-revenuecat-ui-ios-16
+  spm-revenuecat-ui-watchos
+"
+
+for JOB in $JOBS; do
```
Comment (tonidero) [inline .circleci/generate-requested-jobs-config.sh]: So just to check... if there are no jobs passed, this will not work? I guess that's a difference in behavior where running `@RCGitBot please test` would run all tests.

File: .github/workflows/rcgitbot_please_test.yml [hunk 5]
```diff
@@ -125,3 +125,112 @@ jobs:
...
+        id: parse
+        env:
+          COMMENT: ${{ github.event.comment.body }}
+        run: |
+          # Extract everything after "@RCGitBot please test "
+          JOBS_RAW="${COMMENT#@RCGitBot please test }"
+          # Trim whitespace
+          JOBS_RAW=$(echo "$JOBS_RAW" | xargs)
+          if [[ -z "$JOBS_RAW" ]]; then
+            echo "::error::No job name(s) provided"
```
Comment (tonidero) [inline .github/workflows/rcgitbot_please_test.yml:174]: I might be missing something, but this seems to indicate the PR description (that `@RCGitBot please test` should keep working as is) is not accurate? Lmk if I missed something!

**Review by tonidero — COMMENTED** [review]
Looking good, just some questions

File: .circleci/config.yml (hunk 3 above)
Comment (ajpallares) [reply .circleci/config.yml]: Ha... I hadn't thought of this. It's a great suggestion. But I agree that we can and probably should move it to the orb once this is stable and we don't need to iterate on it very often. For now, I'd keep it here because it's only used in purchases-ios for now and to be more agile with iterations if needed. 
But really, great suggestion! 🫶

File: .circleci/config.yml (hunk 1 above)
Comment (ajpallares) [reply .circleci/config.yml]: You're right. My idea was to support both spaces and commas when specifying multiple jobs to run, but it's confusing indeed...
In the end I've simplified to just spaces in https://github.com/RevenueCat/purchases-ios/pull/6607/commits/1562fec342e52628c60b130cc1fb379e63e717a9

File: .github/workflows/rcgitbot_please_test.yml (hunk 5 above)
Comment (ajpallares) [reply .github/workflows/rcgitbot_please_test.yml:174]: You're right it can be confusing. So, the idea is to keep the existing `@RCGitBot please test` behavior while expanding it to allow a list of jobs to run (e.g. `@RCGitBot please test lint api-tests`)

###  `@RCGitBot please test`

Still approves the "Approve all tests" hold job. The condition to trigger this behavior is 
```
github.event.comment.body == '@RCGitBot please test'
```
So `@RCGitBot please test ` (notice the extra blank in the end) would not trigger this code branch.

### `@RCGitBot please test <list_of_jobs>`

This runs all jobs specified in `<list_of_jobs>`. The condition to trigger this code branch is
```
startsWith(github.event.comment.body, '@RCGitBot please test ')
```

So these lines you're mentioning just cover against the edge case of someone typing '@RCGitBot please test ' (with extra blank in the end), which would trigger this code branch but without having specified any actual job. Probably they added the blank by mistake.

Does it make sense?

File: .circleci/generate-requested-jobs-config.sh (hunk 4 above)
Comment (ajpallares) [reply .circleci/generate-requested-jobs-config.sh]: > So just to check... if there are no jobs passed, this will not work? 

Yes, but this script is never called when no jobs are passed. This scripts only runs (see [here](https://github.com/RevenueCat/purchases-ios/pull/6607/changes#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R58-R63)) when triggering the pipeline including a non-empty `requested_jobs` string. 

It could happen that you run the pipeline with `requested_jobs = " "`, and then this would not run anything, true. But a blank string is not a value we're supposed to send for `requested_jobs`.

Now, as explained in [my other comment](https://github.com/RevenueCat/purchases-ios/pull/6607#discussion_r3094584465), `@RCGitBot please test` keeps working as before, by [this code](https://github.com/RevenueCat/purchases-ios/pull/6607/changes#diff-6daba02b846255ec663d4f114b9ba2d454ba846ea0a91ab8bdeaf79fe930f182R7-R13)

File: .circleci/generate-requested-jobs-config.sh (hunk 2 above)
Comment (ajpallares) [reply .circleci/generate-requested-jobs-config.sh]: > Hmm I guess this could potentially be adding secrets unnecessarily to some jobs 🤔. I wonder if this should be some sort of map, to know which contexts are required per job?

You're right it's unnecessary to always pass the secrets. But I did it on purpose to keep things as simple as posible, as I couldn't see any concern for it. Do you think it's worth it to implement the map to only pass the actually-required context to each job? I don't have a strong opinion on this

Comment (ajpallares) [reply .circleci/generate-requested-jobs-config.sh]: > Also, this might be easier to type with typescript/javascript, even though it would require a node machine.

I don't have a strong opinion on this either. And I've tried the JS approach in https://github.com/RevenueCat/purchases-ios/pull/6607/commits/a830d300a15d64c7a4806bdafd131f90c55c6e3e + https://github.com/RevenueCat/purchases-ios/pull/6607/commits/2e0c97fe32f710b8d88ec964fb417c1f2f6bd86e

However, I'm finding that the setup step (the one that creates the actual yml that will end up running) takes considerably longer than using a bare shell script.

* ~JS machine --> The `setup-dynamic-config` takes between 20 and 46s ([46s here](https://app.circleci.com/pipelines/github/RevenueCat/purchases-ios/36632/workflows/ab46affd-56ec-4b3e-9a1e-a173984b581d/jobs/532964))~ Scratch that. It may have been the first few runs. Now It seems consistent at ~10s 😄 
* Base machine --> The `setup-dynamic-config` job takes between 8 and 13s.

Ideally, the `setup` workflow should be as fast as possible, as every pipeline will go through it...

WDYT?

Edit: seems like the speed is exactly the same. So we can have the scripts in JS 👍

Comment (tonidero) [reply .circleci/generate-requested-jobs-config.sh]: I guess it’s not a problem to pass the secrets to all jobs for now, just a bit of leaking things that don’t really need it… i would say it’s a nice to have so we can setup infrastructure more securely but having said that i wouldn’t consider it a blocker.

Also, glad it works fast on JavaScript, thanks for that!!

File: .github/workflows/rcgitbot_please_test.yml (hunk 5 above)
Comment (tonidero) [reply .github/workflows/rcgitbot_please_test.yml:174]: Ohh ok ok. Yeah i didn’t catch that originally but makes sense. Thanks for explaining!

File: .circleci/generate-requested-jobs-config.sh (hunk 2 above)
Comment (ajpallares) [reply .circleci/generate-requested-jobs-config.sh]: Also did the map change in https://github.com/RevenueCat/purchases-ios/pull/6607/commits/8751e3ae1c83328113b51918f5d96ffd532f8cc3 😊

**Review by tonidero — APPROVED** [review]
Looks good! Thanks for doing this!
